Copter: parameterize rotation rate for flip mode

bnsgeyer opened the pull request:
This PR adds a parameter for the user to be able to specify the flip rotation rate for flip mode.  The timeout is calculated based on the flip rate by dividing the flip rate into 360 deg and adding 1 second.  This was tested in SITL with a tradheli frame to ensure it functionally works.  The parameter default rate was set to the rate hard coded which was 400 deg/s.

rmackay9 commented on ArduCopter/Parameters.cpp:
```diff
@@ -1215,6 +1215,16 @@ const AP_Param::GroupInfo ParametersG2::var_info2[] = {
 
+#if MODE_FLIP_ENABLED == ENABLED
+    // @Param: FLIP_RATE
+    // @DisplayName: Rotational Rate for Flip Mode
```
Not a blocker but putting "Flip mode" at the beginning of the description would be good.  So maybe, "Flip Mode Rotational Rate"

bnsgeyer replied:
Done.  Thanks!

rmackay9 commented:
This looks good to me.  I suspect that on real vehicles we should increase the default but that will be easier once people actually try out various rates.

Some thoughts (but not blockers):

- Does ATC_RATE_R/P/Y_MAX still apply when doing flip?  Most users probably leave these at zero but if set they probably apply meaning the user may not get the rate they want and this could lead to a timeout while flipping.
- We could improve the timeout calculation by also using ATC_ACCEL_R/P/Y_MAX

bnsgeyer commented:
@rmackay9 Interestingly enough, the ATC_RATE_RP_MAX does not limit the attitude controller method used to flip the aircraft.

I also find it interesting that you think real vehicles can achieve higher flip rates.  Maybe small quads but I don't think a 450 size heli could achieve 400 deg/s.  Certainly my 600 size heli couldn't.  Trying this in the realflight resulted in some very unexpected results for a heli and is why I am putting in the ability for the user to set the flip rate

bnsgeyer committed:
Copter: parameterize rotation rate for flip mode

bnsgeyer commented:
@rmackay9 @lthall @IamPete1 So I added another commit that is my attempt at making flip a little safer.   I found that the current implementation allows the target attitude and the actual attitude to diverge if the flip rate is not achievable by the aircraft.  I added a feature that limits requested rate if the angle error gets too large.  Looking for input.

IamPete1 commented on ArduCopter/mode_flip.cpp:
```diff
@@ -91,8 +89,12 @@ bool ModeFlip::init(bool ignore_checks)
 {
+    // determine flip rotation rate
+    float flip_rate_cdps = g2.flip_rate_dps.get() * 100.0;
+    float flip_time_out_ms = 1000.0f * (36000.0f / MAX(flip_rate_cdps, 6000.0f) + 1.0);
```
```suggestion
    const float flip_time_out_ms = 1000.0f * (36000.0f / MAX(flip_rate_cdps, 6000.0f));
```
You don't need the +1 because the max guarantees none zero.

Also could be const ;)

bnsgeyer replied:
The +1 was there to give extra time for the aircraft to achieve that rotational rate.  Think of it a fudge factor. The original timeout was at least a second longer than the rotational rate divided into 360.

IamPete1 replied:
Ah, I see, a comment to that effect would be good. Would a multiplier be better than a fixed offset? At 400 deg/s the + 1 is ~ 2.1 times.

bnsgeyer replied:
Made the multiplier 2.0.  The original time out was 2.5 seconds for a 400 deg/s rotation rate.  This decreased that by approximately a half second but I don't think there was much reasoning behind 2.5 seconds.  I think twice the time calculated by the rotation rate is plenty.

IamPete1 commented on ArduCopter/mode_flip.cpp:
```diff
@@ -126,17 +131,24 @@ void ModeFlip::run()
     case FlipState::Roll:
-        // between 45deg ~ -90deg request 400deg/sec roll
-        attitude_control->input_rate_bf_roll_pitch_yaw(FLIP_ROTATION_RATE * roll_dir, 0.0, 0.0);
+        // keep target aircraft from getting too far away from actual aircraft
```
Won't the attitude controller do this stuff for us based on accel limits?

IamPete1 replied:
It seems not. 

However, if we do use the knock down won't we end up timing out?

bnsgeyer replied:
Yeah. I still have to address that. I’ll come up with with something

bnsgeyer replied:
removed this code and saved for a later change.  just want to get adjustable rotation speed in

IamPete1 commented on ArduCopter/mode_flip.cpp:
```diff
@@ -126,17 +131,24 @@ void ModeFlip::run()
+        // keep target aircraft from getting too far away from actual aircraft
+        if (flip_angle_error > 1500 && flip_angle_error <= 4500) {
+            float knock_down = 1.0f - ((float)flip_angle_error - 1500.0f) / 3000.0f;
+            flip_rate_cdps = (ahrs.get_gyro().x + knock_down * (flip_rate_cdps - ahrs.get_gyro().x * 5730)) * roll_dir;
```
```suggestion
            flip_rate_cdps = (ahrs.get_gyro().x + knock_down * (flip_rate_cdps - degrees(ahrs.get_gyro().x) * 100)) * roll_dir;
```
The `degrees` function makes it a little easier to follow.

IamPete1 commented on ArduCopter/mode_flip.cpp:
```diff
@@ -126,17 +131,24 @@ void ModeFlip::run()
+            float knock_down = 1.0f - ((float)flip_angle_error - 1500.0f) / 3000.0f;
+            flip_rate_cdps = (ahrs.get_gyro().x + knock_down * (flip_rate_cdps - ahrs.get_gyro().x * 5730)) * roll_dir;
+        } else if (flip_angle_error > 4500) {
+            flip_rate_cdps = ahrs.get_gyro().x * 5730 * roll_dir;
```
```suggestion
            flip_rate_cdps = degrees(ahrs.get_gyro().x) * 100 * roll_dir;
```

IamPete1 commented on ArduCopter/mode_flip.cpp:
```diff
@@ -89,7 +87,13 @@ bool ModeFlip::init(bool ignore_checks)
 {
-    if (abandon_requested || !motors->armed() || input_is_high_magnitude(*channel_roll) || input_is_high_magnitude(*channel_pitch) || ((millis() - start_time_ms) > FLIP_TIMEOUT_MS)) {
+    // determine flip rotation rate
+    float flip_rate_rads = radians(g2.flip_rate_dps.get());
```
```suggestion
    const float flip_rate_rads = radians(g2.flip_rate_dps.get());
```

IamPete1 commented on ArduCopter/mode_flip.cpp:
```diff
@@ -89,7 +87,13 @@ bool ModeFlip::init(bool ignore_checks)
+    // determine flip rotation rate
+    float flip_rate_rads = radians(g2.flip_rate_dps.get());
+    // determine timeout for flip based on rotation rate, with a minimum of 60deg/s.  Time is doubled to allow for poorly tuned vehicles.
+    float flip_time_out_ms = 1000.0f * (2.0f * 360.0f / MAX(flip_rate_rads, radians(60.0)));
```
```suggestion
    const float flip_time_out_ms = 1000.0f * (2.0f * 360.0f / MAX(flip_rate_rads, radians(60.0)));
```

IamPete1 reviewed (approved):
Looks good, I note that the timeout comes out a bit less by defualt, 1.8 seconds vs 2.5, but I think that is OK.

peterbarker commented on ArduCopter/Parameters.cpp:
```diff
@@ -1185,6 +1185,17 @@ const AP_Param::GroupInfo ParametersG2::var_info2[] = {
+#if MODE_FLIP_ENABLED
+    // @Param: FLIP_RATE
+    // @DisplayName: Flip Mode Rotational Rate
+    // @Description: Rotational Rate for Flip Mode in Deg/s.  Be sure to set a rotational rate that the aircraft can acheive.
```
```suggestion
    // @Description: Rotational Rate for Flip Mode in Deg/s.  Be sure to set a rotational rate that the aircraft can achieve.
```

tridge commented:
**Automated review note - AI-generated (Claude), independently cross-checked by a second model and re-verified against the diff.** Please sanity-check before acting.

_Reviewed at head `8cc9e935f4`._

Verdict: **REQUEST CHANGES**

- **BUG** - [`ArduCopter/mode_flip.cpp:90`](https://github.com/ArduPilot/ardupilot/pull/23578/files) - **The flip timeout mixes degrees with radians and comes out 57.3× too long.**
`const float flip_time_out_ms = 1000.0f * (2.0f * 360.0f / MAX(flip_rate_rads, radians(60.0)));`
The numerator `2 × 360` is degrees of rotation but the divisor is a rate in _radians_ per second, so the quotient is not seconds. At the default `FLIP_RATE=400` deg/s this yields **≈103,000 ms (103 s)** where 1,800 ms was intended; the constant it replaced, `FLIP_TIMEOUT_MS`, was a fixed 2,500 ms. The error factor is exactly 180/π. A 103-second timeout is operationally useless: if the vehicle cannot reach a state transition, Flip keeps issuing its rate and throttle schedule until some other abort path intervenes, so the automatic safety net is effectively gone.
Fix — divide by the deg/s value:
`const float flip_rate_dps = g2.flip_rate_dps.get();`
`const float flip_time_out_ms = 1000.0f * (2.0f * 360.0f / MAX(flip_rate_dps, 60.0f));`

- **ISSUE** - [`ArduCopter/mode_flip.cpp:88`](https://github.com/ArduPilot/ardupilot/pull/23578/files) - **The documented `@Range` is not enforced at runtime**, and the raw parameter is used for the commanded rate while the 60 deg/s floor is applied only to the timeout denominator. So `FLIP_RATE=0` commands no rotation and just waits for the timeout, and a negative value commands rotation opposite the selected direction, so the expected state transition may never occur. Constrain the value once and use that same constrained value for both the command and the timeout.

- **NOTE** - [`ArduCopter/mode_flip.cpp:90`](https://github.com/ArduPilot/ardupilot/pull/23578/files) - **Correction to my own first pass.** I wrote that the timeout is “relaxed regardless” even once the units are fixed. That is wrong. With correct units the crossover is 288 deg/s: 60 deg/s → 12.0 s, 288 → 2.5 s (exactly the old value), 400 → 1.8 s, 1000 → 0.72 s. The timeout is looser only below 288 deg/s and tighter above it, and a 60 deg/s flip legitimately needs more than 2.5 s since a nominal rotation alone takes six seconds. Retained only as a mild design observation.

- **NOTE** - Parameter checks all pass: slot 23 is free in `ParametersG2::var_info2` (1–13 and 17–22 are taken), `FLIP_RATE` is 9 characters against the 16-character limit, and the `AP_Int16`/default 400/range 60–1000/`deg/s` metadata is self-consistent.

- **NOTE** - No test covers the computed timeout. The existing Flip autotests exercise successful flips and explicit RC abandonment, which is why passing CI did not catch the units error.

Full report, including what was checked and found clean: https://uav.tridgell.net/DevCallReviews/DevCallTopic/devcall_pr_reviews.html

tridge commented on ArduCopter/Parameters.cpp:
```diff
@@ -1185,6 +1185,17 @@ const AP_Param::GroupInfo ParametersG2::var_info2[] = {
     AP_GROUPINFO("SURFTRAK_GLSAM", 22, ParametersG2, surf_dist_parameters.glitch_num_samples, AP_SURFACEDISTANCE_GLITCH_NUM_SAMPLES_DEFAULT),
 #endif
 
+#if MODE_FLIP_ENABLED
```
I suspect there will be more flip parameters in the future, maybe create a flip sub-table for flip parameters?

bnsgeyer replied:
done. thanks!

bnsgeyer commented:
Merged! Thanks!
